#Model Algorithm Hyperparameter Tuning:

####Splitting data into Train and Test: Stratified Sampling

In [ ]:
# dropping class label column

y = final_nps_data['Class_label'].values
X = final_nps_data.drop(['Class_label'], axis=1)
X.head(1)

Survey_Time  Time_taken_respond Survey_finished_mobile        Comment  \
0     0.766667           25.388114                    yes  prompt sevice   

  Invitation_Flag   Account_Type              Contract_Type  Claim_Amount  \
0           email  general_agent  future_secure_private_car      117136.0   

  Claim_Status  
0      settled

In [ ]:
# train test split
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, stratify=y)

In [ ]:
print(X_train.shape,y_train.shape)
print(X_test.shape,y_test.shape)

(4696, 9) (4696,)
(1175, 9) (1175,)


####Make Data Model Ready: encoding comments

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.feature_extraction.text import CountVectorizer

#encoding comments into bag of words

vectorizer = CountVectorizer(min_df=10,ngram_range=(1,4), max_features=6500)
vectorizer.fit(X_train['Comment'].values)

X_train_comments_bow = vectorizer.transform(X_train['Comment'].values)
X_test_comments_bow = vectorizer.transform(X_test['Comment'].values)

print(X_train_comments_bow.shape, y_train.shape)

feauture_names= vectorizer.get_feature_names()

(4696, 434) (4696,)


In [ ]:
#encoding comments into tf-idf vectors

vectorizer = TfidfVectorizer(min_df=10,ngram_range=(1,4), max_features=6500)
vectorizer.fit(X_train['Comment'].values)

X_train_comments_tfidf = vectorizer.transform(X_train['Comment'].values)
X_test_comments_tfidf = vectorizer.transform(X_test['Comment'].values)

print(X_train_comments_tfidf.shape, y_train.shape)

feauture_names= vectorizer.get_feature_names()

(4696, 434) (4696,)


####Make Data Model Ready: encoding numerical, categorical features

In [ ]:
#encoding Survey_finished_mobile into one hot encoding

vectorizer = CountVectorizer()
vectorizer.fit(X_train['Survey_finished_mobile'].values)

X_train_Survey_finished_mobile_ohe = vectorizer.transform(X_train['Survey_finished_mobile'].values)
X_test_Survey_finished_mobile_ohe = vectorizer.transform(X_test['Survey_finished_mobile'].values)

print(X_train_Survey_finished_mobile_ohe.shape, y_train.shape)

(4696, 2) (4696,)


In [ ]:
#encoding Invitation_Flag into one hot encoding

vectorizer = CountVectorizer()
vectorizer.fit(X_train['Invitation_Flag'].values)

X_train_Invitation_Flag_ohe = vectorizer.transform(X_train['Invitation_Flag'].values)
X_test_Invitation_Flag_ohe = vectorizer.transform(X_test['Invitation_Flag'].values)

print(X_train_Invitation_Flag_ohe.shape, y_train.shape)

(4696, 2) (4696,)


In [ ]:
#encoding Account_Type into one hot encoding

vectorizer = CountVectorizer()
vectorizer.fit(X_train['Account_Type'].values)

X_train_Account_Type_ohe = vectorizer.transform(X_train['Account_Type'].values)
X_test_Account_Type_ohe = vectorizer.transform(X_test['Account_Type'].values)

print(X_train_Account_Type_ohe.shape, y_train.shape)

(4696, 11) (4696,)


In [ ]:
#encoding Contract_Type into one hot encoding

vectorizer = CountVectorizer()
vectorizer.fit(X_train['Contract_Type'].values)

X_train_Contract_Type_ohe = vectorizer.transform(X_train['Contract_Type'].values)
X_test_Contract_Type_ohe = vectorizer.transform(X_test['Contract_Type'].values)

print(X_train_Contract_Type_ohe.shape, y_train.shape)

(4696, 36) (4696,)


In [ ]:
#encoding Claim_Status into one hot encoding

vectorizer = CountVectorizer()
vectorizer.fit(X_train['Claim_Status'].values)

X_train_Claim_Status_ohe = vectorizer.transform(X_train['Claim_Status'].values)
X_test_Claim_Status_ohe = vectorizer.transform(X_test['Claim_Status'].values)

print(X_train_Claim_Status_ohe.shape, y_train.shape)

(4696, 10) (4696,)


In [ ]:
# encoding numerical feauture Survey_Time

from sklearn.preprocessing import Normalizer
normalizer = Normalizer()
normalizer.fit(X_train['Survey_Time'].values.reshape(-1,1))

X_train_Survey_Time_norm = normalizer.transform(X_train['Survey_Time'].values.reshape(-1,1))
X_test_Survey_Time_norm = normalizer.transform(X_test['Survey_Time'].values.reshape(-1,1))

print(X_train_Survey_Time_norm.shape, y_train.shape)

(4696, 1) (4696,)


In [ ]:
# encoding numerical feauture Time_taken_respond

from sklearn.preprocessing import Normalizer
normalizer = Normalizer()
normalizer.fit(X_train['Time_taken_respond'].values.reshape(-1,1))

X_train_Time_taken_respond_norm = normalizer.transform(X_train['Time_taken_respond'].values.reshape(-1,1))
X_test_Time_taken_respond_norm = normalizer.transform(X_test['Time_taken_respond'].values.reshape(-1,1))

print(X_train_Time_taken_respond_norm.shape, y_train.shape)

(4696, 1) (4696,)


In [ ]:
# encoding numerical feauture Claim_Amount

from sklearn.preprocessing import Normalizer
normalizer = Normalizer()
normalizer.fit(X_train['Claim_Amount'].values.reshape(-1,1))

X_train_Claim_Amount_norm = normalizer.transform(X_train['Claim_Amount'].values.reshape(-1,1))
X_test_Claim_Amount_norm = normalizer.transform(X_test['Claim_Amount'].values.reshape(-1,1))

print(X_train_Claim_Amount_norm.shape, y_train.shape)

(4696, 1) (4696,)


In [ ]:
# stacking all features horizantally bow_comments

from scipy.sparse import hstack
X_tr = hstack((X_train_comments_bow,X_train_Survey_finished_mobile_ohe,X_train_Invitation_Flag_ohe,X_train_Account_Type_ohe,X_train_Contract_Type_ohe,X_train_Claim_Status_ohe,X_train_Survey_Time_norm,X_train_Time_taken_respond_norm,X_train_Claim_Amount_norm)).tocsr()
X_te = hstack((X_test_comments_bow,X_test_Survey_finished_mobile_ohe,X_test_Invitation_Flag_ohe,X_test_Account_Type_ohe,X_test_Contract_Type_ohe,X_test_Claim_Status_ohe,X_test_Survey_Time_norm,X_test_Time_taken_respond_norm,X_test_Claim_Amount_norm)).tocsr()

print(X_tr.shape, y_train.shape)
print(X_te.shape, y_test.shape)

(4696, 498) (4696,)
(1175, 498) (1175,)


In [ ]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV
hyper_param = {'min_samples_split':[5, 10, 100, 500],'max_depth': [1, 3, 10, 30]}
model = GridSearchCV(DecisionTreeClassifier(), hyper_param,scoring='roc_auc', cv=3)
model.fit(X_tr, y_train)

GridSearchCV(cv=3, estimator=DecisionTreeClassifier(),
             param_grid={'max_depth': [1, 3, 10, 30],
                         'min_samples_split': [5, 10, 100, 500]},
             scoring='roc_auc')

In [ ]:
best_RF_model = model.best_estimator_
best_RF_model

DecisionTreeClassifier(max_depth=30, min_samples_split=500)

In [ ]:
best_RF_model.fit(X_tr, y_train)
y_train_pred_RF = best_RF_model.predict(X_tr)
y_test_pred_RF = best_RF_model.predict(X_te)

In [ ]:
from sklearn.metrics import roc_auc_score

print('train f1 score', roc_auc_score(y_train, y_train_pred_RF))
print('test f1 score', roc_auc_score(y_test, y_test_pred_RF))

train f1 score 0.8030405495442674
test f1 score 0.7931327129677703


In [ ]:
# stacking all features horizantally bow_comments

from scipy.sparse import hstack
X_tr = hstack((X_train_comments_tfidf,X_train_Survey_finished_mobile_ohe,X_train_Invitation_Flag_ohe,X_train_Account_Type_ohe,X_train_Contract_Type_ohe,X_train_Claim_Status_ohe,X_train_Survey_Time_norm,X_train_Time_taken_respond_norm,X_train_Claim_Amount_norm)).tocsr()
X_te = hstack((X_test_comments_tfidf,X_test_Survey_finished_mobile_ohe,X_test_Invitation_Flag_ohe,X_test_Account_Type_ohe,X_test_Contract_Type_ohe,X_test_Claim_Status_ohe,X_test_Survey_Time_norm,X_test_Time_taken_respond_norm,X_test_Claim_Amount_norm)).tocsr()

print(X_tr.shape, y_train.shape)
print(X_te.shape, y_test.shape)

(4696, 498) (4696,)
(1175, 498) (1175,)


In [ ]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV
hyper_param = {'min_samples_split':[5, 10, 100, 500],'max_depth': [1, 3, 10, 30]}
model_tfidf = GridSearchCV(DecisionTreeClassifier(), hyper_param,scoring='roc_auc', cv=3)
model_tfidf.fit(X_tr, y_train)

GridSearchCV(cv=3, estimator=DecisionTreeClassifier(),
             param_grid={'max_depth': [1, 3, 10, 30],
                         'min_samples_split': [5, 10, 100, 500]},
             scoring='roc_auc')

In [ ]:
best_RF_tfidf_model = model.best_estimator_
best_RF_tfidf_model

DecisionTreeClassifier(max_depth=30, min_samples_split=500)

In [ ]:
best_RF_model.fit(X_tr, y_train)
y_train_pred_RF_tfidf = best_RF_tfidf_model.predict(X_tr)
y_test_pred_RF_tfidf = best_RF_tfidf_model.predict(X_te)

In [ ]:
from sklearn.metrics import roc_auc_score

print('train f1 score', roc_auc_score(y_train, y_train_pred_RF_tfidf))
print('test f1 score', roc_auc_score(y_test, y_test_pred_RF_tfidf))

train f1 score 0.8244296679281489
test f1 score 0.7760094252965375


# Conclusion:

Let's tabulate the results.

In [ ]:
# tabulate results
from prettytable import PrettyTable
x = PrettyTable()

x.field_names = ["Vectorizer", "Model", "AUC Score"]
x.add_rows(
    [
        ["BOW", 'DecisionTreeClassifier', roc_auc_score(y_test, y_test_pred_RF)],
        ["TF_IDF",'DecisionTreeClassifier',roc_auc_score(y_test, y_test_pred_RF_tfidf)],
    ]
)

print(x)

+------------+------------------------+--------------------+
| Vectorizer |         Model          |     AUC Score      |
+------------+------------------------+--------------------+
|    BOW     | DecisionTreeClassifier | 0.7931327129677703 |
|   TF_IDF   | DecisionTreeClassifier | 0.7760094252965375 |
+------------+------------------------+--------------------+


*   Here we can see auc score from random forest model is 79%.
*   As number of feedbacks are less we can improve the performance by collecting more feedbacks.
*   Also performance of model can be improved by using complex model.



